### ISSUE TITLE
USB drive support and SAS drives

### ISSUE THREAD

**Issue by 007revad** [wallacebrf/Synology-SMART-test-scheduler#7]
I needed to test a SATA SSD connected via a USB dock today and I thought you should add USB drive support :o)

If you want to add support for USB SATA SSD and USB SATA HDD I can do that for you?

I've actually started on it (just for fun). I've appended the USB drives to disk_list1_exploded or disk_list2_exploded (depending on whichever array is not empty).
```
root@SENNA:~# /volume1/web/synology_smart/synology_SMART_control_test.sh
Disk /dev/sata1: 931.5 GiB, 1000204886016 bytes, 1953525168 sectors
Disk /dev/sata2: 167.7 GiB, 180045766656 bytes, 351651888 sectors
Disk /dev/sata3: 465.8 GiB, 500107862016 bytes, 976773168 sectors
Disk /dev/usb1: 960 MiB, 1006632960 bytes, 1966080 sectors
Disk /dev/usb2: 7.5 GiB, 8004304896 bytes, 15633408 sectors
Disk /dev/usb3: 59.6 GiB, 64023257088 bytes, 125045424 sectors
```
I had to change the 4 instances `smartctl -a -d ata` to `smartctl -a -d sata` for smartctl to work with USB SATA SSDs. 
```
Not yet time to scan Synology Drive Slot: 1 [Main Unit]. Next scan scheduled for Fri Feb  7 10:38:00 AM AEDT 2025
Synology Drive Slot: 1 [Main Unit]
Disk: /dev/usb3
Model: STT_FTM64GX25H
Serial: P595418-redacted
User Capacity:    64,023,257,088 bytes [64.0 GB]

Testing is already in progress.
Percent complete: 100%
```
It looks like my SSD smart test is stuck at 100% and won't end... but I was testing this 20 year old SSD because files on it are becoming corrupted so I'm not surprised.
```
Self-test routine in progress...
00% of test remaining.
```

I just need to fix the USB drives currently showing as internal drives:
```
Synology Drive Slot: 1 [Main Unit]
Synology Drive Slot: 2 [Main Unit]
```
so it can show:
```
Synology USB Port: 1
Synology USB Port: 2
```

<br>
I was wondering if smartctl works with SAS drives and apparently `smartctl -a` does but it provides less info for SAS drives. To get the full info with SAS drives you need to use `smartctl -x`. I have no idea if `-d sata` or `-d ata` work with SAS drives.

**Comment by wallacebrf** [comment]
Go ahead and keep testing and updating the code. Once you have something that is complete, make a pull request and I will merge it into the main branch

**Comment by wallacebrf** [comment]
I will make sure to add detail to the credits that the USB/SAS support etc is thanks to your assistance 

Do you have SAS drives? I do not do i cannot test that myself. 

I do have an external USB "spinning rust" hard drive and I have an external flash SSD USB drive I can use to test your updates one the pull request is merged into the main branch

**Comment by 007revad** [comment]
Unfortunately I don't have a SAS drive, or a NAS that supports SAS drives. But I do know SAS drives use different device names:
- /dev/sas1, /dev/sas2 etc in newer Synology NAS.
- /dev/hda, /dev/hdb etc in older Synology NAS.

I've changed lines 213 and 214 to find SATA and SAS drives.

Here's the output with 2 USB thumb drives (in a USB hub) and a SATA SSD in a USB dock. I just need to change it to skip USB thumb drives.
```
root@SENNA:~# /volume1/web/synology_smart/synology_SMART_control.sh
Not yet time to scan Synology Drive Slot: 1 [Main Unit]. Next scan scheduled for                                                                                                                                                              Fri Feb  7 10:38:00 AM AEDT 2025
Synology Drive Slot: 1 [Main Unit]
Disk: /dev/sata1
Model: WD10JUCT-63CYNY0
Serial: redacted
User Capacity:    1,000,204,886,016 bytes [1.00 TB]

Testing not active. Test Result: PASSED

#################################################################


Not yet time to scan Synology Drive Slot: 2 [Main Unit]. Next scan scheduled for                                                                                                                                                              Fri Feb  7 10:38:00 AM AEDT 2025
Synology Drive Slot: 2 [Main Unit]
Disk: /dev/sata2
Model: SSDSC2BF180A4H
Serial: redacted
User Capacity:    180,045,766,656 bytes [180 GB]

Testing not active. Test Result: PASSED

#################################################################


Not yet time to scan Synology Drive Slot: 1 [Expansion Unit 1]. Next scan schedu                                                                                                                                                             led for Fri Feb  7 10:38:00 AM AEDT 2025
Synology Drive Slot: 1 [Expansion Unit 1]
Disk: /dev/sata3
Model: ST3500312CS
Serial: redacted
User Capacity:    500,107,862,016 bytes [500 GB]

Testing not active. Test Result: PASSED

#################################################################


23:02:01

Attention!!
Synology USB Port: 2
Disk: /dev/usb1

SMART is not enabled. Unable to start Extended testing.


23:02:01

Attention!!
Synology USB Port: 2
Disk: /dev/usb2

SMART is not enabled. Unable to start Extended testing.


Not yet time to scan Synology USB Port: 1. Next scan scheduled for Fri Feb  7 10                                                                                                                                                             :38:00 AM AEDT 2025
Synology USB Port: 1
Disk: /dev/usb3
Model: STT_FTM64GX25H
Serial: redacted
User Capacity:    64,023,257,088 bytes [64.0 GB]

Testing is already in progress.
Percent complete: 100%

#################################################################


```

**Comment by 007revad** [comment]
I'm changing the 4 lines that are echoing blank lines to the shell if the 4 "if, else, fi" that check if the file already exists.
```
if [ -r "$file" ]; then
	echo ""
else
	echo "foo bar" > "$file"
fi
```
to:
```
if [[ ! -r "$file" ]]; then
	echo "foo bar" > "$file"
fi
```

<br>
I've tested the changes on my Asustor. It's USB drives are /dev/sd* and I haven't found a way to skip USB thumb drives (but it's not really a problem).

**Comment by wallacebrf** [comment]
rolled pull request into main branch and made a correction to the ```disk_scan_status.txt``` file data saving.

**Comment by wallacebrf** [comment]
connected 12TB WD elements disk i have been meaning to perform SMART testing on anyways to my test DS920. 

even though the disk is NOT visible in storage manager, it was successfully recognized by the script. 

i have successfully started a manual long test on it. will let it finish. 

```
Not yet time to scan Synology USB Port: 1. Next scan scheduled for Mon Nov 11 06:34:00 PM CST 2024
Synology USB Port: 1
Disk: /dev/usb1
Model: WD120EDAZ-11F3RA0
Serial: redacted
User Capacity:    12,000,138,625,024 bytes [12.0 TB]

Testing is already in progress.
Percent complete: 10%
```

what i might do, seeing as how someone might added and removed USB drives to facilitate testing, i might change the code so that the /log/YYYY-MM-DD_usb1 file name ends with the disk's serial number. this way if they test more than one disk in 24-hours, the old logs do not get over-written....

i might update the code so ALL disk file names include the disk serial number, that way there does not need to be added logic to check if the disk is a USB before making the log file....

**Comment by 007revad** [comment]
Good idea with the serial number. That will also help if someone replaces a drive too.

You should also check if the log_dir exists, and create it if it doesn't. I just a clean install with the following and got errors when the script tried to write to /volume1/web/synology_smart/log/disk_scan_status.txt
```
/volume1/web/synology_smart/config
/volume1/web/synology_smart/config/smart_control_config.txt
/volume1/web/synology_smart/synology_SMART_control.sh
```

I'm currently working on USB drives in older Synology NAS that use /dev/sdx for USB drives instead of /dev/usb1. In an 8 bay Synology it is showing as:
```
Synology Drive Slot: 17 [Main Unit]
Disk: /dev/sdq
```

**Comment by wallacebrf** [comment]
good idea on the checking and verification of expected directories. 

tonight i am done for now. i will update the code tomorrow to add the serial numbers to the log file file-names, and i will add the logic to check for and created needed directories. might as well also add logic to ensure the directories are writable in-case the permissions are not working or something. 

i like to have scripts and programs try to account for as many error-producing scenarios as possible so the script can handle the issue gracefully.

**Comment by wallacebrf** [comment]
With your permission I woul like to eventually update the smart to influxdb script I have to include the new code for USB support and the ability to get drive slot etc.

**Comment by 007revad** [comment]
No problem. Thanks for asking first.
